text/0000-normalized-modules.md

```diff
@@ -0,0 +1,535 @@
+of modules allowed in each bucket.
+
+The following buckets are proposed:
+
+* `-components` - components and their corresponding templates
+* `-data` - including models, adapters, and serializers
+* `-helpers` - helpers
+* `-initializers` - application and instance initializers
+* `-routes` - routes, templates, and routable components
+* `-services` - services
+
+### Files for an example application
+
+Given the base and alternative _file_ naming patterns, the example application
+could be restructured using the following files and directories:
+
+```
+modules
+  -routes
+    posts
+      new
+        -components
+          new-post
+            component.js
+            template.hbs
+        route.js
+        template.hbs
+      -components
+        post-editor
+          -helpers
+            calculate-post-title.helper.js
+          post-editor-button
+            component.js
+            template.hbs
+          component.js
+          template.hbs
+        post-viewer
+          component.js
+          template.hbs
+      -helpers
+        capitalize.helper.js
+        titleize.helper.js
+      route.js
+      template.hbs
+  -components
+    list-paginator
+      component.js
+      template.hbs
+  -data
+    comments
+      adapter.js
+      model.js
+      serializer.js
+    posts
+      adapter.js
+      model.js
+      serializer.js
+  -helpers
+    capitalize.helper.js
+  -initializers
+    auth.instance-initializer.js
+    i18n.initializer.js
+  -services
+    auth.service.js
+  app.js
+  router.js
+  template.hbs
+```
+
+This example shows that we can improve the ergonomics of naming _files_ in
+Ember CLI projects, while normalizing everything into a single _module_
+naming structure.
+
+## Normalized module:file map
+
+The Ember CLI resolver will need to maintain a map of normalized module
+names to file locations on disk. This map will provide a single place for
+module lookups by the resolver.
+
+Construction of this map will allow for detection of any module name conflicts.
+In the case of any conflicts, Ember CLI should throw an exception.
+
+## Addon modules
+
+An addon's modules will no longer be "mixed in" to a consuming application
+directly. Instead, the addon's modules will be accessible at a namespace.
+Addons will be able to publish a default namespace (e.g. `ember-power-select`
+might choose `power-select`), which consuming applications will be able to
+alias to avoid conflicts.
+
+Private modules will exist alongside public modules in addons within a
+`-private` namespace. This namespace will provide a strong signal to consumers
+that private modules should not be accessed.
+
+### Namespaced components and helpers
+
+A `::` separator will be used to namespace components and helpers from
+templates.
+
+If an addon namespaced as `magic-ui` has a root-level `select/component`
+module, then that component can be rendered as `{{magic-ui::select}}`.
```

Comment:
@rwjblue I left a longer note below. This may not be a breaking change for curlies, but as things stand curlies would be relying on (assumed to be deprecated) lookup patterns, and there is no path forward from curlies to angle bracket components that use this pattern.